Add quantity to inventory units and split on demand

priyank-gupta opened the pull request:
# Optimising Shipments and Inventory Units.
1. Creates a single `inventory_unit` per `state`, per `line_item`, per `stock_location` with a `quantity` field indicating inventory units in that state.
2. Adds a `return_quantity` field in `ReturnAuthorization` indicating the number of units to be returned/exchanged.
3. Split inventory unit to extract the portion/quantity that needs to be returned. This might be further split depending upon stock availability at the time of generating return.
4. Changes prioritizer and adjuster to maximise the number of `on_hand` items from available stocks in multiple `stock_locations`
5. Change `splitters/weight.rb` to use **Worst Fit** algo.

## Prioritizer - Example Scenario
**Stock Availability:**
* Stock Location 1: 50,  Not Backorderable
* Stock Location 2: 150, Backorderable
* Stock Location 3: 200, Backorderable

**Location Priority:**
* Stock Location 1 > Stock Location 2 > Stock Location 3 (Based on creation)

**Stock Requested: 10**
* Old : 10 from Stock Location 1
* New : 10 from Stock Location 1

**Stock Requested: 100**
* Old : 50 from Stock Location 1, 50 from Stock Location 2
* New : 50 from Stock Location 1, 50 from Stock Location 2

**Stock Requested: 300**
* Old : 50 from Stock Location 1, 150 from Stock Location 2, 100 Backordered from Stock Location 2
* New : 50 from Stock Location 1, 150 from Stock Location 2, 100 from Stock Location 3

**Stock Requested: 450**
* Old : 50 from Stock Location 1, 150 from Stock Location 2, 250 Backordered from Stock Location 2
* New : 50 from Stock Location 1, 150 from Stock Location 2, 200 from Stock Location 3, 50 Backordered from Stock Location 2

## Weight Splitter - Example Scenario and Benchmarks
#### Threshold Weight: 150

**Item Weights:** 49, 73, 70, 52, 87, 66, 63, 44, 57, 33, 89, 36, 64, 81, 22, 95, 54, 90, 58, 53
* Old Split:
  * 33.0, 22.0, 95.0 | 90.0, 36.0 | 89.0, 44.0 | 87.0, 49.0 | 81.0, 52.0 | 73.0, 53.0 | 70.0, 54.0 | 66.0, 57.0 | 64.0, 58.0 | 63.0
* New Split:
  * 95.0, 53.0 | 90.0, 54.0 | 89.0, 57.0 | 87.0, 58.0 | 81.0, 66.0 | 73.0, 70.0 | 64.0, 63.0 | 52.0, 49.0, 44.0 | 36.0, 33.0, 22.0


**Item Weights:** 74, 80, 18, 59, 67, 29, 33, 52, 53, 48, 61, 46, 60, 82, 67, 50, 99, 99, 36, 24
* Old Split:
  * 18.0, 99.0, 24.0 | 99.0, 29.0 | 82.0, 33.0 | 80.0, 36.0 | 74.0, 46.0 | 67.0, 48.0 | 67.0, 50.0 | 61.0, 52.0 | 60.0, 53.0 | 59.0
* New Split:
  * 99.0, 50.0 | 99.0, 48.0 | 82.0, 61.0 | 80.0, 67.0 | 74.0, 67.0 | 60.0, 59.0, 18.0 | 53.0, 52.0, 29.0 | 46.0, 36.0, 33.0, 24.0


**Item Weights:** 79, 26, 73, 59, 24, 34, 47, 18, 44, 19, 25, 54, 10, 63, 59, 30, 62, 15, 87, 43
* Old Split:
  * 18.0, 19.0, 10.0, 15.0, 87.0 | 79.0, 25.0, 24.0 | 73.0, 30.0, 26.0 | 63.0, 43.0, 34.0 | 62.0, 44.0 | 59.0, 47.0 | 59.0, 54.0
* New Split:
  * 87.0, 59.0 | 79.0, 62.0 | 73.0, 63.0, 10.0 | 59.0, 54.0, 34.0 | 47.0, 44.0, 43.0, 15.0 | 30.0, 26.0, 25.0, 24.0, 19.0, 18.0

### With benchmarks and 40 items:
**Item Weights:** 52, 60, 37, 31, 4, 8, 52, 56, 91, 92, 26, 2, 50, 0, 38, 98, 49, 67, 55, 6, 27, 27, 98, 39, 77, 30, 4, 48, 99, 54, 66, 95, 53, 10, 92, 94, 63, 39, 0, 43
* Old Split
  * 4.0, 8.0, 2.0, 0.0, 6.0, 4.0, 99.0, 10.0, 0.0 | 98.0, 26.0 | 98.0, 27.0 | 95.0, 27.0 | 94.0, 30.0 | 92.0, 31.0 | 92.0, 37.0 | 91.0, 38.0 | 77.0, 39.0 | 67.0, 43.0, 39.0 | 66.0, 48.0 | 63.0, 49.0 | 60.0, 50.0 | 56.0, 52.0|55.0, 52.0|54.0, 53.0
  * user: **0.290000**, system: **0.000000**, total: **0.290000**, real: **0.296323**
* New Split
  * 99.0, 48.0 | 98.0, 50.0 | 98.0, 49.0 | 95.0, 52.0 | 94.0, 52.0 | 92.0, 54.0 | 92.0, 53.0 | 91.0, 55.0 | 77.0, 67.0 | 66.0, 63.0 | 60.0, 56.0, 27.0 | 43.0, 39.0, 39.0, 27.0 | 38.0, 37.0, 31.0, 30.0 | 26.0, 10.0, 8.0, 6.0, 4.0, 4.0, 2.0, 0.0, 0.0
  * user: **0.030000**, system: **0.000000**, total: **0.030000**, real: **0.025941**

**Item Weights:** 66, 10, 2, 91, 53, 81, 89, 89, 56, 63, 44, 43, 33, 33, 84, 20, 35, 78, 98, 50, 63, 34, 68, 65, 3, 97, 84, 56, 5, 70, 19, 90, 54, 17, 16, 66, 95, 46, 70, 59
* Old Split:
  * 10.0, 2.0, 98.0, 3.0, 5.0, 16.0 | 97.0, 19.0, 17.0 | 95.0, 33.0, 20.0 | 91.0, 33.0 | 90.0, 34.0 | 89.0, 35.0 | 89.0, 43.0 | 84.0, 44.0 | 84.0, 46.0 | 81.0, 50.0 | 78.0, 53.0 | 70.0, 54.0 | 70.0, 56.0 | 68.0, 56.0 | 66.0, 59.0 | 66.0, 63.0 | 65.0, 63.0
  * user: **0.310000**, system: **0.000000**, total: **0.310000**, real: **0.315362**
* New Split:
  * 98.0, 46.0 | 97.0, 50.0 | 95.0, 53.0 | 91.0, 54.0 | 90.0, 56.0 | 89.0, 59.0 | 89.0, 56.0 | 84.0, 66.0 | 84.0, 65.0 | 81.0, 66.0 | 78.0, 70.0 | 70.0, 68.0, 10.0 | 63.0, 63.0, 16.0, 2.0 | 44.0, 43.0, 35.0, 17.0, 5.0 | 34.0, 33.0, 33.0, 20.0, 19.0, 3.0
  * user: **0.030000**, system: **0.000000**, total: **0.030000**, real: **0.023759**

**Item Weights:** 99, 19, 8, 34, 93, 66, 19, 4, 37, 47, 78, 27, 39, 16, 74, 50, 24, 92, 27, 84, 7, 33, 25, 12, 95, 72, 26, 83, 13, 41, 55, 98, 32, 93, 95, 55, 55, 64, 39, 92
* Old Split:
  * 99.0, 8.0, 4.0, 7.0, 12.0, 13.0 | 98.0, 19.0, 16.0 | 95.0, 24.0, 19.0 | 95.0, 26.0, 25.0 | 93.0, 27.0, 27.0 | 93.0, 32.0 | 92.0, 33.0 | 92.0, 34.0 | 84.0, 37.0 | 83.0, 39.0 | 78.0, 39.0 | 74.0, 41.0 | 72.0, 47.0 | 66.0, 50.0 | 64.0, 55.0 | 55.0, 55.0
  * user: **0.280000**, system: **0.000000**, total: **0.280000**, real: **0.276268**
* New Split:
  * 99.0, 37.0, 12.0 | 98.0, 39.0, 7.0 | 95.0, 41.0, 8.0 | 95.0, 39.0, 13.0 | 93.0, 50.0 | 93.0, 47.0 | 92.0, 55.0 | 92.0, 55.0 | 84.0, 55.0, 4.0 | 83.0, 64.0 | 78.0, 66.0 | 74.0, 72.0 | 34.0, 33.0, 32.0, 27.0, 16.0 | 27.0, 26.0, 25.0, 24.0, 19.0, 19.0
  * user: **0.020000**, system: **0.000000**, total: **0.020000**, real: **0.024061**

Commit:
Only split inventory units if needed.

Commit:
Add scopes for singular inventory units, update js

Commit:
Add validations to return quantity.

Commit:
restore original adjuster and priortizer

Use && instead of and

Commit:
Optimise the spree weight splitter

Commit:
Update weight splitter to split inventory units if needed.

Commit:
Rename the content item split method

Commit:
Minor refactoring and name changes.

Commit:
Do not split on return into singular inventory on returns.

Commit:
Refactor weight splitter to be more self contained.

Commit:
Use inventory unit's quantity to restock item.

Fix frontend specs.

Stock availability validator fix

Fix shipment specs

Commit:
Fix specs where quantity was not mocked. Test partial fulfillment

Commit:
Use sum(&:quantity) instead to handle in memory inventory units.

Commit:
Fix default refund amount calculation, when adjustments are present.

Commit:
Build inventory units from hash instead of builder.

Fix indentation

Use === for equality comparisons

Commit:
Make prioritizer work for distributing stock items across multiple stock location.

Commit:
Revisit multi inventory for order inventory and other touched portions, fix specs

Commit:
fix reimbursement object for exchange

Commit:
return item can now have many exchange inventory units

minor code rafctor

mauazua reviewed (commented):
hi @priyank-gupta could you please address hound issues? Besides that, looks fine to me.

Mafi88 commented on core/lib/tasks/exchanges.rake:
```diff
@@ -3,22 +3,20 @@ namespace :exchanges do
     # Determine that a return item has already been deemed unreturned and therefore charged
     # by the fact that its exchange inventory unit has popped off to a different order
-    unreturned_return_items.select! { |ri| ri.inventory_unit.order_id == ri.exchange_inventory_unit.order_id }
+    unreturned_return_items.select! { |ri| ri.exchange_inventory_units.where(order_id: ri.inventory_unit.order_id).exists? }
```
```
unreturned_return_items.select! { |ri| ri.exchange_inventory_units.exists?(order_id: ri.inventory_unit.order_id) }
```

Commit:
reload inventory units in order inventory and fix build

priyank-gupta commented on core/app/models/spree/inventory_unit.rb:
```diff
@@ -79,6 +104,22 @@ def included_tax_total
       line_item.included_tax_total * percentage_of_line_item
     end
 
+    def empty?
```
@Mafi88 Right, inventory unit quantity will not be 0 in any case. Infact, `singular?` and `empty?` methods are not used anywhere. Removing them.

Commit:
Code refactors for inventory unit optimisation

fix hound issues

fix hound issues

fix hound issues

fix hound issues

fix hound issues

minor typo fix

refactor migration

priyank-gupta commented:
@Mafi88 I am done with the suggested changes. Also rebased with latest master.

damianlegawiec commented on core/db/migrate/20170119122701_add_original_return_item_id_to_spree_inventory_units.rb:
```diff
@@ -0,0 +1,29 @@
+
+    Spree::InventoryUnit.reset_column_information
+
+    Spree::ReturnItem.where.not(exchange_inventory_unit_id: nil).each do |return_item|
```
@priyank-gupta please use `.find_each` instead of each, there could be a huge number of records like this

damianlegawiec commented on core/db/migrate/20170119122701_add_original_return_item_id_to_spree_inventory_units.rb:
```diff
@@ -0,0 +1,29 @@
+
+    Spree::InventoryUnit.reset_column_information
+
+    Spree::InventoryUnit.where.not(original_return_item_id: nil).each do |inventory_unit|
```
@priyank-gupta same here ;)

priyank-gupta commented:
@damianlegawiec Done with the suggested changes.
